blog post: Julia 1.7 Highlights (#1419)
JuliaLang/www.julialang.org: blog/2021/10/julia-1.7-highlights.md

```julia-repl
julia> using Random; Random.seed!("seed")
ERROR: MethodError: no method matching seed!(::String)
Closest candidates are:
  seed!() at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.6/Random/src/RNGs.jl:362
...
```

Some editors have support for opening files by clicking paths in the terminal but this does not work in cases like this. In v1.7, this has been fixed to instead print paths that are valid locally:

```julia-repl
julia> using Random; Random.seed!("seed")
ERROR: MethodError: no method matching seed!(::String)
Closest candidates are:
  seed!() at ~/Downloads/julia/share/julia/stdlib/v1.7/Random/src/RNGs.jl:387
...
```

The example above was for a `MethodError` but the same improvement also applies to stacktraces.

## Inference improvements

*Shuhei Kadowaki*

This release comes with many type inference improvements.
With these improvements, Julia v1.7 will more "smartly" infer types of your program and improve the performance for free !

Mostly notably, v1.7 is able to propagate type constraints that can be derived from `isa` and `===` conditions _inter-procedurally_ (i.e. across any function calls).

> Certain Julia programs are written in a way that it changes the behavior depending on runtime type, and such programs may run much faster by the inferrability gain of this improvement. For example, now there is no inferrability difference between `x === nothing` and `isnothing(x)` (and so you no longer need to remember [this performance tip](https://docs.julialang.org/en/v1.6/manual/performance-tips/#Checking-for-equality-with-a-singleton)):

Reviewer: ```suggestion
Certain Julia programs are written in a way that their behavior changes depending on runtime types, and such programs may run much faster by the inferrability gain of this improvement. For example, now there is no inferrability difference between `x === nothing` and `isnothing(x)` (and so you no longer need to remember [this performance tip](https://docs.julialang.org/en/v1.6/manual/performance-tips/#Checking-for-equality-with-a-singleton)):
```

```julia-repl
julia> code_typed((Union{Nothing,Int},); optimize=false) do x
           return isnothing(x) ? 0 : x
       end |> first
```
```diff
--- v1.6
+++ v1.7
@@ -1,6 +1,6 @@
 CodeInfo(
 1 ─ %1 = Main.isnothing(x)::Bool
 └──      goto #3 if not %1
 2 ─      return 0
-3 ─      return x
-) => Union{Nothing, Int64}
+3 ─      return x::Int64
+) => Int64
```

Of course, this inter-procedural constraint propagation works for arbitrary generic functions:
```julia-repl
julia> ispositive(a) = isa(a, Number) && a > 0;
julia> code_typed((Union{Nothing,Int},); optimize=false) do x
           return ispositive(x) ? x : 0
       end |> first
```
```diff
--- v1.6
+++ v1.7
@@ -1,6 +1,6 @@
 CodeInfo(
 1 ─ %1 = Main.ispositive(x)::Bool
 └──      goto #3 if not %1
-2 ─      return x
+2 ─      return x::Int64
 3 ─      return 0
-) => Union{Nothing, Int64}
+) => Int64
```

> Another remarkable improvement is more eager constant-propagations.

Reviewer: ```suggestion
Another remarkable improvement is more eager constant propagation.
```

> Julia v1.7 can substitute more runtime computations with pre-computed constants, and eliminate dead code by resolving conditional branches at compile-time.

Reviewer: ```suggestion
Julia 1.7 can substitute more runtime computations with pre-computed constants, and eliminate dead code by resolving conditional branches at compile time.
```

> As an example, on v1.7, computations of special-functions can be fully folded at compile-time:

Reviewer: ```suggestion
As an example, in 1.7, computations of special functions can be fully folded at compile time:
```

```julia-repl
julia> code_typed((Int,)) do n

>            n + sin(sum(sincos(42))) # no runtime computation of `sum(sincos(42))` on v1.7 !

Reviewer: ```suggestion
           n + sin(sum(sincos(42))) # no runtime computation of `sum(sincos(42))` in 1.7 !
```

Reviewer 2: ```suggestion
           n + sin(sum(sincos(42))) # no runtime computation of `sum(sincos(42))` in 1.7!
```

       end |> first
```
```diff
--- v1.6
+++ v1.7
@@ -1,32 +1,5 @@
 CodeInfo(
-1 ─ %1  = Base.muladd_float(0.16933292771007588, 2.7557313707070068e-6, -0.0001984126982985795)::Float64
-│   %2  = Base.muladd_float(0.16933292771007588, %1, 0.00833333333332249)::Float64
-│   %3  = Base.muladd_float(0.16933292771007588, 1.58969099521155e-10, -2.5050760253406863e-8)::Float64
-│   ... many runtime computations ...
-│   %27 = invoke Main.sin(%26::Float64)::Float64
-│   %28 = Base.sitofp(Float64, n)::Float64
-│   %29 = Base.add_float(%28, %27)::Float64
-└──       return %29
+1 ─ %1 = Base.sitofp(Float64, n)::Float64
+│   %2 = Base.add_float(%1, -0.9678422808766897)::Float64
+└──      return %2
 ) => Float64
```

> For those interested, here is the list of specific PRs that implement main inference improvements of this release:

Reviewer: ```suggestion
For those interested, here is the list of specific PRs that implement the main inference improvements of this release:
```

- inter-procedural conditional constraint propagation ([#38905](https://github.com/JuliaLang/julia/pull/38905))

> - constant-propagation for union-split callsite ([#39305](https://github.com/JuliaLang/julia/pull/39305))
> - constant-propagation for `invoke` callsite ([#41383](https://github.com/JuliaLang/julia/pull/41383))

Reviewer: ```suggestion
- constant propagation for union-split callsite ([#39305](https://github.com/JuliaLang/julia/pull/39305))
- constant propagation for `invoke` callsite ([#41383](https://github.com/JuliaLang/julia/pull/41383))
```

- more conditional constraint propagation ([#39936](https://github.com/JuliaLang/julia/pull/39936), [#40832](https://github.com/JuliaLang/julia/pull/40832))

> Actually these inference improvements were motivated by the need for [JET.jl](https://github.com/aviatesk/JET.jl), a static analyzer for Julia, that is powered by Julia compiler's type inference implementation.

Reviewer: ```suggestion
These inference improvements were initially motivated by the needs of [JET.jl](https://github.com/aviatesk/JET.jl), a static analyzer for Julia, that is powered by Julia compiler's type inference implementation.
```

> These inference improvements of v1.7 allow JET to analyze your program more correctly and faster –

Reviewer: ```suggestion
These inference improvements in 1.7 allow JET to analyze your program more correctly and faster –
```

as a simple measurement, [when analyzing JET itself](https://gist.github.com/aviatesk/e2ffa4bfaee60f939ef4b65449fa394b),

> JET took `90` seconds to reported `93` false positive errors on v1.6,
> but on v1.7 and higher, JET can finish the analysis within `40` seconds and the number of false positives is reduced to `27`,

Reviewer: ```suggestion
JET took `90` seconds to reported `93` false positive errors in 1.6,
but in 1.7 and higher, JET can finish the analysis within `40` seconds and the number of false positives is reduced to `27`,
```

A simple example of how this syntax can be used:

```julia-repl
julia> nt = (a=1, b=2)
(a = 1, b = 2)

julia> (; a, b) = nt
(a = 1, b = 2)

julia> a
1

julia> b
2
```

`(; a, b) = nt` is simply equivalent to `a = nt.a; b = nt.b`, so it can be used not just for `NamedTuple`s, but for destructuring the properties of any object and will also work with `DataFrame`s for example. It can also be used inside function arguments just like regular destructuring. A `myreal` function which just extracts the real part of a complex number could now be written as follows:

```julia-repl
julia> myreal((; re)::Complex) = re
myreal (generic function with 1 method)

julia> myreal(2 + 3im)
2
```

This syntax is also compatible with type annotations (note that these don't play any role in dispatch when used inside function arguments though):

```julia-repl
julia> let
           (; a::UInt, b::Float64) = (a=1, b=2)
           b = 5
           a, b
       end
(0x0000000000000001, 5.0)
```

## Conclusion

> Please enjoy the release, and as always [let us know](https://github.com/JuliaLang/julia/issues) if you encounter any problems or have any suggestions. We hope to be back in about four months to report on even more progress in version 1.8!

Reviewer: Is four months a bit too optimistic? :sweat_smile:

Author: Hehe, well, we do hope for four months, don't we? :P

Reviewer 3: Maybe "a few months"? :smile:

+++
mintoclevel = 2
maxtoclevel = 2
title = "Julia 1.7 highlights"
+++

> After 4 betas and 2 release candidates, Julia version 1.7 has finally been released. We would like to thank all the contributors to this release and all the testers that helped with finding regressions and issues in the pre-releases. Without you, this release would not have been possible.

Reviewer 4: 3 release candidates.

Author: So far ;)

We knew we might want to re-evaluate that choice at some point, but it didn't seem
particularly urgent until Chet Hega [pointed out](https://github.com/JuliaLang/julia/pull/34852)
that by changing algorithms we could not only get a significant speed-up, but also
make random number streams *reproducible in multi-threaded programs*.

Mersenne Twister is famous for having an exceptionally long period, but that requires
a correspondingly large state, and is not truly necessary for any practical application.
We also used thread-local RNG states for thread safety, which has a bit of overhead and made
random streams task-schedule-dependent.
Chet's proposal took advantage of the much smaller state of the [Xoshiro256 family](https://prng.di.unimi.it/)
of RNGs to put a state in every *task*, and fork it on each task creation.
That makes random numbers depend only on the task spawning structure of a program, and not
on the parallel execution schedule.
It took a bit of debate for all of us to be comfortable spending precious bytes of Task
objects this way, but we are huge fans of reproducibility and so the proposal was eventually
adopted.

## New atomic support

> Jameson Nash

Reviewer 5: ```suggestion
*Jameson Nash*
```
